controllers/machine_controller_node.go

```diff
@@ -0,0 +1,150 @@
+	capierrors "sigs.k8s.io/cluster-api/errors"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	ErrNodeNotFound = errors.New("cannot find node with matching ProviderID")
+)
+
+func (r *MachineReconciler) reconcileNode(ctx context.Context, cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
+	logger := r.Log.WithValues("machine", machine.Name, "namespace", machine.Namespace)
+	// Check that the Machine hasn't been deleted or in the process.
+	if !machine.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	logger = logger.WithValues("cluster", cluster.Name)
+
+	// Check that the Machine has a valid ProviderID.
+	if machine.Spec.ProviderID == nil || *machine.Spec.ProviderID == "" {
+		logger.Info("Machine doesn't have a valid ProviderID yet")
+		return nil
+	}
+
+	providerID, err := noderefutil.NewProviderID(*machine.Spec.ProviderID)
+	if err != nil {
+		return err
+	}
+
+	remoteClient, err := r.Tracker.GetClient(ctx, util.ObjectKey(cluster))
+	if err != nil {
+		return err
+	}
+
+	// Even if Status.NodeRef exists, continue to do the following checks to make sure Node is healthy
+	node, err := r.getNode(remoteClient, providerID)
+	if err != nil {
+		if err == ErrNodeNotFound {
+			// While a NodeRef is set in the status, getting ErrNodeNotFound error means the node is deleted.
+			// If Status.NodeRef is not set before, node still can be in the provisioning state.
+			if machine.Status.NodeRef != nil {
+				conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeNotFoundReason, clusterv1.ConditionSeverityError, "")
+			}
+			return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 20 * time.Second},
+				"cannot assign NodeRef to Machine %q in namespace %q, no matching Node", machine.Name, machine.Namespace)
+		}
+		logger.Error(err, "Failed to assign NodeRef")
+		r.recorder.Event(machine, apicorev1.EventTypeWarning, "FailedSetNodeRef", err.Error())
+		return err
+	}
+
+	// Set the Machine NodeRef.
+	if machine.Status.NodeRef == nil {
+		machine.Status.NodeRef = &apicorev1.ObjectReference{
+			Kind:       node.Kind,
+			APIVersion: node.APIVersion,
+			Name:       node.Name,
+			UID:        node.UID,
+		}
+		logger.Info("Set Machine's NodeRef", "noderef", machine.Status.NodeRef.Name)
+		r.recorder.Event(machine, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", machine.Status.NodeRef.Name)
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeProvisioningReason, clusterv1.ConditionSeverityWarning, "")
+		return nil
+	}
+
+	// Do the remaining node health checks, then set the node health to true if all checks pass.
+	if err := hasHealthyNodeConditions(node); err != nil {
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeConditionsFailedReason, clusterv1.ConditionSeverityWarning, "Unhealthy conditions: %s", err.Error())
```

Comment:
The resulting message is kind of verbose
```
NAME                                                                   READY  SEVERITY  REASON                SINCE  MESSAGE
Cluster/test1                                                          True                                   4m20s
├─ControlPlane - KubeadmControlPlane/test1-control-plane               True                                   4m20s
│ └─Machine/test1-control-plane-dtxqk                                  True                                   4m21s
│   │           ├─NodeHealthy                                          False  Warning   NodeConditionsFailed  4m20s  Unhealthy conditions: Condition Ready is false on Node: test1-control-plane-dtxqk
```
What about 
- dropping `Unhealthy conditions:`, because this is redundant given that there is already  `NodeConditionsFailed` 
- dropping on Node: test1-control-plane-dtxqk given that the conditions is already linked to a machine/node; if not, at least we should avoid to repeat the node name for each failing condition